Remotion/Web/Core/Context/HttpContextStorageProviderNetCore.cs

```diff
@@ -0,0 +1,208 @@
+using Remotion.ServiceLocation;
+using Remotion.Utilities;
+
+namespace Remotion.Web.Context
+{
+#if !NETFRAMEWORK
+  [ImplementationFor(typeof(ISafeContextStorageProvider), Position = 0)]
+  public class HttpContextStorageProvider : ISafeContextStorageProvider, ISafeContextBoundaryStrategy
+  {
+    private struct HashTableAccessor
+    {
+      private class WrappedHashtable
+      {
+        public HttpContext Context { get; }
+        public Hashtable Data { get; }
+
+        public WrappedHashtable (HttpContext context, Hashtable data)
+        {
+          Context = context;
+          Data = data;
+        }
+      }
+
+      private const string c_dataDictionaryKey = "Rm.DD";
+
+      private readonly ThreadLocal<WeakReference<WrappedHashtable?>> _dataTable;
+
+      public HashTableAccessor (bool _)
+      {
+        _dataTable = new(() => new WeakReference<WrappedHashtable?>(null));
+      }
+
+      public bool TryGetCachedHashtable (HttpContext httpContext, [MaybeNullWhen(false)] out Hashtable hashtable)
+      {
+        if (TryGetCachedWrappedHashtable(httpContext, out var cachedWrappedHashtable))
+        {
+          hashtable = cachedWrappedHashtable.Data;
+          return true;
+        }
+
+        hashtable = null;
+        return false;
+      }
+
+      public Hashtable GetOrCreateHashtable (HttpContext context)
+      {
+        if (TryGetCachedHashtable(context, out var hashtable))
+        {
+          return hashtable;
+        }
+
+        return SetHashtable(context, new Hashtable());
+      }
+
+      public Hashtable? RemoveHashtable (HttpContext context)
+      {
+        if (TryGetCachedWrappedHashtable(context, out var cachedHashtable))
+        {
+          context.Items.Remove(c_dataDictionaryKey);
+          _dataTable.Value!.SetTarget(null!);
+        }
+
+        return cachedHashtable?.Data;
+      }
+
+      public Hashtable SetHashtable (HttpContext context, Hashtable hashtable)
+      {
+        var wrappedHashtable = new WrappedHashtable(context, hashtable);
+        context.Items[c_dataDictionaryKey] = wrappedHashtable;
+        _dataTable.Value!.SetTarget(wrappedHashtable);
+        return hashtable;
+      }
+
+      private bool TryGetCachedWrappedHashtable (HttpContext httpContext, [MaybeNullWhen(false)] out WrappedHashtable hashtable)
+      {
+        if (_dataTable.Value!.TryGetTarget(out var target) && target.Context == httpContext)
+        {
+          hashtable = target;
+          return true;
+        }
+
+        var wrappedHashtable = (WrappedHashtable?)httpContext.Items[c_dataDictionaryKey];
+        hashtable = wrappedHashtable;
+
+        if (wrappedHashtable != null)
+        {
+          _dataTable.Value!.SetTarget(wrappedHashtable);
+#pragma warning disable CS8762 // Parameter must have a non-null value when exiting in some condition.
+          return true;
+#pragma warning restore CS8762 // Parameter must have a non-null value when exiting in some condition.
```

Comment:
Shouldn't it be possible to avoid those by using `hashtable != null` instead of `wrappedHashtable`?